src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs

```diff
@@ -0,0 +1,399 @@
+    }
+
+    fn connection_pool_ttl(&self) -> Duration {
+        self.connection_pool_ttl.clone()
+    }
+
+    fn connection_pool_ttl_stagger(&self) -> Duration {
+        self.connection_pool_ttl_stagger.clone()
+    }
+
+    fn connect_timeout(&self) -> Duration {
+        self.connect_timeout.clone()
+    }
+
+    fn tcp_user_timeout(&self) -> Duration {
+        self.tcp_user_timout.clone()
+    }
+}
+
+impl<N> PostgresTimestampOracle<N>
+where
+    N: GenericNowFn<Timestamp> + 'static,
+{
+    /// Open a Postgres [`TimestampOracle`] instance with `config`, for the
+    /// timeline named `timeline`. `next` generates new timestamps when invoked.
+    /// Timestamps that are returned are made durable and will never retract.
+    pub(crate) async fn open(
+        config: PostgresTimestampOracleConfig,
+        timeline: String,
+        initially: Timestamp,
+        next: N,
+    ) -> Result<Self, anyhow::Error> {
+        info!(url = ?config.url, "opening PostgresTimestampOracle");
+
+        let metrics = Arc::clone(&config.metrics);
+
+        let postgres_client = PostgresClient::open(config.into())?;
+
+        let client = postgres_client.get_connection().await?;
+
+        // WIP: Do we need this for the pg oracle? This is originally from
+        // persist.
+        //
+        // The `consensus` table creates and deletes rows at a high frequency, generating many
+        // tombstoned rows. If Cockroach's GC interval is set high (the default is 25h) and
+        // these tombstones accumulate, scanning over the table will take increasingly and
+        // prohibitively long.
+        //
+        // See: https://github.com/MaterializeInc/materialize/issues/13975
+        // See: https://www.cockroachlabs.com/docs/stable/configure-zone.html#variables
+        client
+            .batch_execute(&format!(
+                "{} {}",
+                SCHEMA, "ALTER TABLE timestamp_oracle CONFIGURE ZONE USING gc.ttlseconds = 600;",
+            ))
+            .await?;
+
+        let mut oracle = PostgresTimestampOracle {
+            timeline,
+            next,
+            postgres_client: Arc::new(postgres_client),
+            metrics,
+        };
+
+        // Create a row for our timeline, if it doesn't exist. The `apply_write`
+        // call below expects the row to be present. If we didn't have this here
+        // we would always need CHECK EXISTS calls or something in
+        // `apply_write`, making it more complicated, so we only do it once
+        // here, on initialization.
+        let q = r#"
+            INSERT INTO timestamp_oracle (timeline, read_ts, write_ts)
+                VALUES ($1, $2, $3)
+                ON CONFLICT (timeline) DO NOTHING;
+        "#;
+        let statement = client
+            .prepare_cached(q)
+            .await
+            .expect("todo: add internal retries");
+        let _ = client
+            .execute(&statement, &[&oracle.timeline, &0i64, &0i64])
```

Comment:
I'd be more comfortable if we used `initially` here instead of the `0`s. Any reason not to do that?

Reply:
done :ok_hand: